Aggregated pull request

tpetricek opened the pull request:
I ended up doing things in way too many branches, so this is an aggregated PR with everything.

To review the important changes, please see the previous PRs, which include comments about the changes and also show a nice diff.
- [Make RProvider fail with expected error message on Mac](https://github.com/BlueMountainCapital/FSharpRProvider/pull/136)
- [Make RProvider work on Mac](https://github.com/BlueMountainCapital/FSharpRProvider/pull/137)

tpetricek committed:
Make RProvider fail with expected error message on Mac (1/4 for #133)

 - Waiting until parent process quits does not work, so this is
   done in a simple async loop in `RProvider.Server.exe`
 - Named events do not work, so instead, `RProvider` creates a temp file
   which is deleted by the server once it is ready (and the provider
   waits for this using `FileSystemWatcher`).
 - Tweak logging so that it can be enabled using `RPROVIDER_LOG`
   environment variable
 - I also did minor naming tweaks and refactoring of long functions
   in the source files that I touched
 - Make `fsproj` files compatible with Xamarin Studio

tpetricek committed:
Make R provider work on Mac

Currently, Xamarin studio runs as 32 bit process and R runs only as 64 bit
process, so to make this work, you have to compile 64 bit version of mono.
However, once that's done, this should work. There are two tricks:

 - You need to create `~/.rprovider.conf` pointing to your 64 bit mono
   (the file should contain e.g. `MONO64=/usr/local/mono64/bin/mono`)
 - You need to have R in PATH so that we can run `R --print-home` to
   get R location.

The main changes are:

 - In `RProvider.Server.exe` - this exposes the access to R via remoting.
   On Mono, this did not work giving the evil "C stack size exceeded" error.
   This is usually related to threading, so I changed this to access the R
   engine through a simple event loop running on the main thread (this is
   the only thing that worked on Mac) - this is for IntelliSense only and
   I think it is safer choice for Windows too.

 - In `RInteropClient.fs`, we need to find location of 64bit mono
   and if that fails, we report a nice error (through IntelliSense).

 - In `RInit.fs`, I added a new approach to resolving R location, which
   is only used on Mac and calls `R --print-home`.

 - Also, `RTypeBuilder.fs` returned a lazy enumerable from a function that
   was wrapped in a lock, which is obviously unintended, so I fixed that
   (afaik, it didn't cause any deadlocks, so this should be fine)

tpetricek committed:
Merge remote-tracking branch 'RussellPolitzky/master'

tpetricek committed:
Prefer the 'R' key when it exists, otherwise scan all

tpetricek committed:
Merge changes and recover timeout on starting server

tpetricek committed:
Merge changes (searching for R_Home)

tpetricek committed:
Use ILRepack to make RProvider.Server.exe completely stand-alone

As suggested by @forki, this change uses ILRepack to merge FSharp.Core.dll (and RProvider.Runtime.dll)
with RProvider.Server.exe. This means that the executable is standalone and we do not have to worry
about mismatching versions of FSharp.Core (or NuGet/VS getting confused because there is FSharp.Core
in the RProvider package - that's no longer needed).

tpetricek commented:
Added [Use ILRepack to make RProvider.Server.exe completely stand-alone](https://github.com/BlueMountainCapital/FSharpRProvider/commit/9dbe52a46d7ae216c99c7ce855be96229085403f) with the following comment:

As suggested by @forki, this change uses ILRepack to merge `FSharp.Core.dll` (and `RProvider.Runtime.dll`) with `RProvider.Server.exe`. This means that the executable is standalone and we do not have to worry about mismatching versions of `FSharp.Core.dll` (or NuGet/VS getting confused because there is `FSharp.Core.dll` in the RProvider package - that's no longer needed).

PS: Installing a VM to test this...

tpetricek committed:
Update R provider to use Paket & try fix Travis

The only notable thing about this is that I had to move RProvider.DesignTime to a
separate directory (a good thing anyway) so that it can have its own `paket.references`
file (specifying additional dependencies). This is now also using github reference
to get the type provider starter pack files.

tpetricek committed:
Skip tests on Travis (needs R) and remove F# types from RInteropServer (fix ilmerge)

tpetricek committed:
Alpha release and support Paket folder structure (#133)

tpetricek committed:
Fix remaining ILMerge issues (avoid F# types in server)

tpetricek committed:
Reference ILRepack via NuGet but disable it (does not work on Mono)

tpetricek committed:
Fix and release alpha NuGet package

forki commented:
you should add paket.exe to gitignore and rebase interactive in order to completely get rid of the binary.

forki commented:
you should remove the RestorePackages target from the build.fsx

forki commented on build.fsx:
```diff
@@ -135,7 +163,7 @@ Target "NuGet" (fun _ ->
             Tags = tags
             OutputPath = "bin"
-            ToolPath = nugetPath
+            ToolPath = "nuget.exe"
```
just remove the toolpath parameter completely. FAKE will discover nuget.exe in the packages folder.

forki commented on nuget/RProvider.nuspec:
```diff
@@ -30,8 +34,8 @@
-    <file src="..\nuget\FSharp.Core.dll" target="lib\net40"/>
-    <file src="..\nuget\FSharp.Core.optdata" target="lib\net40"/>
-    <file src="..\nuget\FSharp.Core.sigdata" target="lib\net40"/>
+    <file src="..\tools\FSharp.Core.dll" target="lib\net40"/> 
```
why don't you use FSharp.Core nuget package like https://github.com/fsprojects/Paket/blob/master/paket.dependencies#L11 ?

forki commented on nuget/publish.cmd:
```diff
@@ -1 +1 @@
-@for %%f in (..\bin\*.nupkg) do @..\.nuget\NuGet.exe push %%f
\ No newline at end of file
+@for %%f in (..\bin\*.nupkg) do @..\packages\NuGet.CommandLine\tools\NuGet.exe push %%f
```
if you setup the release process like in http://fsprojects.github.io/ProjectScaffold/release-process.html you don't need this extra publish.cmd

forki commented on paket-files/fsprojects/FSharp.TypeProviders.StarterPack/src/DebugProvidedTypes.fs:
```diff
@@ -0,0 +1,519 @@
+﻿// --------------------------------------------------------------------------------------
```
Consider to gitignore the paket-files folder. This will be restored everytime you build.

forki commented on paket.dependencies:
```diff
@@ -0,0 +1,20 @@
+github fsprojects/FSharp.TypeProviders.StarterPack src/ProvidedTypes.fs
+github fsprojects/FSharp.TypeProviders.StarterPack src/DebugProvidedTypes.fs
+
+nuget FsCheck 0.9.2.0
```
you should consider to remove the version restrictionss here. This is probably to restrictive.

usually we use the paket.lock file to pin the versions. If you keep the restrictions here, then `paket update`will always pull these versions.

forki commented on paket.dependencies:
```diff
@@ -0,0 +1,20 @@
+nuget FSharp.Compiler.Service 0.0.44
+nuget FSharp.Data 2.0.7
+nuget FSharp.Formatting 2.4.8
+nuget Microsoft.AspNet.Razor 2.0.30506.0
```
if you run `paket simplify` then this dependency will go away since it's only indirect.

forki commented on paket.lock:
```diff
@@ -0,0 +1,32 @@
+    xunit.runners (1.9.2)
+    Zlib.Portable (1.10.0)
+GITHUB
+  remote: fsprojects/FSharp.TypeProviders.StarterPack
```
It's so cool to see these things happening ;-)

tpetricek committed:
Expose Server via an interface and remove dead code

The server needs to be exposed as an interface, because F# Interactive on Mono
doesn't seem to like DLL files referencing EXE files (so the client cannot
reference the server directly).

This also removes the (never used) option for local execution, because this would
need testing (if we need it, it can be easily recovered) & one not used function.
Added documentation as a bonus :-).

tpetricek committed:
Add forgotten interface file

tpetricek committed:
Fix RValue serialization and add tests

tpetricek committed:
Wait for file deletion in a loop (works on Mono)

tpetricek committed:
Add workaround for Printing on Mono/Mac

For some reason, the standard implementation of Print hangs on
Mono/Mac. This workaround implements an alternative version (used
only on *nix systems) that prints the output to a file. This is more
expensive, but it's only used when printing output in FSI.

tpetricek committed:
Adding diagnostics and Mac pages

tpetricek committed:
Finish writing new documentaiton pages

tpetricek committed:
Adding Release notes

tpetricek committed:
Update version, fix RValue serialization and test

tpetricek committed:
Remove Paket binary

tpetricek committed:
Address paket-files, FSharp.Core and NuGet commands

tpetricek committed:
Finish Paket conversion improvements

tpetricek commented:
@forki Thanks a lot for the feedback - all the things you mentioned should be corrected now!

@hmansell @dcharbon OK - So, as far as I can tell, this now does all that it should be doing! Sorry for making such a big change that is going to be difficult to review. If you could have a look at this, that would be fantastic.

forki commented:
:shipit:

tpetricek committed:
Downgrade R.NET reference and new release

tpetricek commented:
I had an issue with the latest R.NET on Mac, so I downgraded the referenced version. The problem is logged [on CodePlex](https://rdotnet.codeplex.com/workitem/154)

tpetricek committed:
Fix typos in the docs

dcharbon commented:
The changes make sense to me, I haven't tried to download and test, yet.

tpetricek commented:
@dcharbon Thanks! I'm [outsourcing testing to Twitter :-)](https://twitter.com/tomaspetricek/status/555052851781988352). But I'll finish setting up a clean VS2013 VM and try there too.

tpetricek commented:
Tested this on a clean VM machine with just VS 2013 and it works fine (no missing FSharp.Core issue :-))

tpetricek committed:
Make base functionality available without open RProvider.``base`` (fix #2)

I implemented this while doing other changes and forgot about it :-).
This is done simply by providing another "R" type, directly inside
the RProvider namespace with a copy of all "base" functions (if you
open both, you'll get the more recently opened one, which is fine as
they're the same).

tpetricek committed:
Merge branch 'mainline' of github.com:BlueMountainCapital/FSharpRProvider into mainline

tpetricek committed:
Released as non-alpha version

tpetricek commented:
OK then, I'm merging this, but I'll keep the branch around in case @hmansell has time for a review.

hmansell commented:
I took a quick look.  I don't see anything obviously troubling from the diffs... I like the use of IL merging to simplify the dependency management of the server.  

If it works then I am happy!

forki commented:
:shipit:

tpetricek commented:
Sadly, the IL merging isn't working at the moment, so I commented that out. (At least, I did another fix that tells NuGet not to add FSharp.Core.dll as a reference, which was sometimes causing issues.)

The latest version of the `ILRepack` tool fails on our assembly - but it is an active GitHub project, so I'll see if we can get that fixed. (Older version worked on Windows, but not on Mac...)
